My First NFT Guide Update (#838)
aptos-labs/developer-docs: apps/nextra/pages/en/build/guides/your-first-nft.mdx

```tsx filename="index.ts"
const createCollectionTransaction = await aptos.createCollectionTransaction({
  creator: alice,
  description: "This is an example collection.",
  name: "Example Collection",
  uri: "aptos.dev",
});

const committedTxn = await aptos.signAndSubmitTransaction({
  signer: alice,
  transaction: createCollectionTransaction,
});
await aptos.waitForTransaction({ transactionHash: committedTxn.hash });
```

<Callout type="info">
Always wait for the transaction to complete using `waitForTransaction` before proceeding. This ensures the collection is ready before you try minting.
</Callout>

### Mint a Digital Asset

With the collection created, we can now mint a digital asset (an NFT) for the collection. This involves providing details like the name, description, and a URI (often linking to metadata like images).

```tsx filename="index.ts"
const mintTokenTransaction = await aptos.mintDigitalAssetTransaction({
  creator: alice,
  collection: "Example Collection",
  description: "This is an example digital asset.",
  name: "Example Asset",

>   uri: "aptos.dev/asset",

Reviewer: ```suggestion
  uri: "https://aptos.dev/asset.png",
```

const digitalAssetAddress = aliceDigitalAssets[0].token_data_id;

const transferTransaction = await aptos.transferDigitalAssetTransaction({
  sender: alice,
  digitalAssetAddress,
  recipient: bob.accountAddress,
});

const transferTxn = await aptos.signAndSubmitTransaction({
  signer: alice,
  transaction: transferTransaction,
});
await aptos.waitForTransaction({ transactionHash: transferTxn.hash });
```

After completion, the asset should now appear in Bob's account.

### Verify the Balances

Finally, we check both Alice's and Bob's accounts to ensure that Alice no longer has the asset and Bob now has it.

```tsx filename="index.ts"
const aliceDigitalAssetsAfter = await aptos.getOwnedDigitalAssets({ ownerAddress: alice.accountAddress });
const bobDigitalAssetsAfter = await aptos.getOwnedDigitalAssets({ ownerAddress: bob.accountAddress });

console.log(`Alice's digital asset balance: ${aliceDigitalAssetsAfter.length}`);
console.log(`Bob's digital asset balance: ${bobDigitalAssetsAfter.length}`);
```
</Steps>

## Running An Example

### Getting Started

<Steps>
### Prerequisites

- **Node.js**: Make sure you have Node.js (version 16 or later) and npm installed.

> - **Aptos CLI (Optional)**: Having the [Aptos CLI](https://aptos.dev/cli-tools/aptos-cli-tool/install-aptos-cli) can be helpful if you want to inspect on-chain data directly, but it's not required for this tutorial.

Reviewer: Nit: rm

Author: removed the prerequisites section 👍

### Getting Started

<Steps>
### Prerequisites

- **Node.js**: Make sure you have Node.js (version 16 or later) and npm installed.
- **Aptos CLI (Optional)**: Having the [Aptos CLI](https://aptos.dev/cli-tools/aptos-cli-tool/install-aptos-cli) can be helpful if you want to inspect on-chain data directly, but it's not required for this tutorial.

### Set up Your Project

Create a new directory for your project and initialize a Node.js project:

```bash filename="Terminal"
mkdir aptos-digital-asset-tutorial
cd aptos-digital-asset-tutorial
npm init -y
```

This will create a `package.json` file, allowing you to install dependencies and run scripts.

### Install Dependencies

You will need the Aptos TypeScript SDK and `dotenv` to manage environment variables:

```bash filename="Terminal"
npm install @aptos-labs/ts-sdk dotenv
npm install --save-dev @types/node
```

### Create tsconfig.json

1. Create a new TypeScript configuration file:

<Tabs items={["Mac/Linux", "Windows"]}>
  <Tabs.Tab>
    ```bash filename="Terminal"
    touch tsconfig.json
    ```
  </Tabs.Tab>
  <Tabs.Tab>
    ```bash filename="Terminal"
    type nul > tsconfig.json
    ```
  </Tabs.Tab>
</Tabs>

> 2. Open the `tsconfig.json` file and add the following configuration:

Reviewer: Can we skip this? I think we should assume that the user knows how to make a new file on their machine. Ideally it just looks like "Create a tsconfig file with the following:"

```json filename="tsconfig.json"
{
  "compilerOptions": {
    "target": "es2020",
    "module": "commonjs",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true,
    "types": ["node"],
    "lib": ["es2020"]
  }
}
```

<Callout type="warning">
Remember to save your file after adding the configuration!

> </Callout>

Reviewer: This also seems extraneous

This configuration ensures TypeScript properly recognizes Node.js types and provides appropriate type checking.

### Configure Environment Variables

1. Create a `.env` file by running:

<Tabs items={["Mac/Linux", "Windows"]}>
  <Tabs.Tab>
    ```bash filename="Terminal"
    touch .env
    ```
  </Tabs.Tab>
  <Tabs.Tab>
    ```bash filename="Terminal"
    type nul > .env

>     ```

Reviewer: Can we skip this? I think we should assume that the user knows how to make a new file on their machine. Ideally it just looks like "Create a `.env` file with the following:"

  </Tabs.Tab>
</Tabs>

2. Open `.env` to edit it.
3. Update `.env` to have the following line:

```bash filename="Terminal"
APTOS_NETWORK=devnet
```

<Callout type="warning">
Remember to save your `.env` file after adding the configuration!

> </Callout>

Reviewer: Nit: Extraneous

Author: they gone

```bash filename="Terminal"
=== Step 7: Transaction hashes for explorer ===

Collection creation transaction: 0x8c5d2a4ce32d76349bfb4f3830740c1c103399e8cbc31d6e2c7a871c88e6ad48
Mint transaction: 0x673d2cbb9fef468fe41f271c0fcf20872e9fa79afb6a2000368394000071b02e
Transfer transaction: 0x3a1e99d6fd3f8e7e962c311f3dfd92c11e468da5b6084123b8f7e0248a37ffa7

You can view these transactions on the Aptos Explorer:
https://explorer.aptoslabs.com/?network=devnet
```

You can view these transactions on the Aptos Explorer:

1. Copy the transaction hash from your console
2. Visit [Aptos Explorer](https://explorer.aptoslabs.com/?network=devnet)
3. Make sure you're on the correct network (Devnet)
4. Paste the transaction hash in the search bar
5. View the details of your transaction, including:
   - The sender and recipient addresses
   - The exact time the transaction was processed
   - Gas fees paid
   - The digital asset that was transferred

This is a great way to verify your transactions and understand how they're recorded on the blockchain.
</Steps>

> **Congrats!**

Reviewer: Lol maybe let's remove, looks strange

---
title: "Your First NFT"
---

import { Callout, Steps } from 'nextra/components'

# Your First NFT

This tutorial will guide you through the process of using the Aptos TypeScript SDK (`@aptos-labs/ts-sdk`) to create a new digital asset (often referred to as an NFT) on Aptos. By the end of this tutorial, you will know how to:

1. Create a collection of digital assets (NFTs).
2. Mint a new digital asset (NFT) within that collection.
3. Transfer the digital asset (NFT) between accounts.
4. Verify the digital asset's (NFT's) movement by checking the updated balances.

<Callout type="info">

> This tutorial assumes you are comfortable with using the command line, have Node.js and npm installed, and understand basic JavaScript/TypeScript concepts. If you need more info, check out Node.js Documentation or the Aptos TypeScript SDK documentation.

Reviewer: Should we add links for `Node.js Documentation` and `Aptos TypeScript SDK documentation`

Author: We should. Done [x]

> This tutorial assumes you are comfortable with using the command line, have Node.js and npm installed, and understand basic JavaScript/TypeScript concepts. If you need more info, check out Node.js Documentation or the Aptos TypeScript SDK documentation.

Reviewer: ```suggestion
This tutorial assumes you are comfortable with using the [Aptos CLI](TODOLINK), have Node.js and npm installed, and understand basic JavaScript/TypeScript concepts. If you need more info, check out Node.js Documentation or the Aptos TypeScript SDK documentation.
```

Author: Yep. Done [x]

</Callout>

## Walking Through The Code

> Below is the step-by-step explanation of how to create, transfer, and interact with a digital asset on-chain. This goes through how the example code (shown in full at the end) does. To skip to just getting the code running, see [**Running An Example**](#running-an-example).

Reviewer: > This goes through how the example code (shown in full at the end) does. 

Hmm this sentence doesn't make grammatical sense to me

Author: Should be "We'll go through how the example code (shown in full at the end) does it". Done [x]

### Key Steps

- **Collection Creation**: A "collection" is a grouping of digital assets. We'll create a new collection under Alice's account.
- **Minting a Digital Asset**: Once the collection is in place, we'll mint a new digital asset (NFT) under that collection.

> - **Transferring a Digital Asset**: Finally, we'll transfer the newly minted asset from Alice to Bob.

Reviewer: Nit: remove these steps are mentioned above already, repetitive to have both

Author: Done

### Code Walkthrough

<Steps>

> ### Setup the Client

Reviewer: Nit: newline between 32 and 33

Author: Fixed in most recent commit

We import and configure the `Aptos` client from the SDK to connect to the specified network:

```tsx filename="index.ts"
const APTOS_NETWORK = NetworkToNetworkName[process.env.APTOS_NETWORK] || Network.DEVNET;
const config = new AptosConfig({ network: APTOS_NETWORK });
const aptos = new Aptos(config);
```

This `aptos` object allows us to interact with the Aptos blockchain (funding accounts, creating assets, submitting transactions, etc.).

### Create and Fund Accounts

> We generate two accounts, Alice and Bob. On devnet, we can easily fund them with test coins so they can cover the gas fees for transactions.

Reviewer: ```suggestion
We generate two accounts, Alice and Bob. On devnet, we can easily fund them with test APT
```

Author: Fixed in my next commit [x]. (Your suggestion was missing a ., otherwise would have directly committed suggestion)

```tsx filename="index.ts"
const alice = Account.generate();
const bob = Account.generate();

await aptos.fundAccount({ accountAddress: alice.accountAddress, amount: INITIAL_BALANCE });
await aptos.fundAccount({ accountAddress: bob.accountAddress, amount: INITIAL_BALANCE });
```

### Create a Collection

We create a collection in Alice's account. A collection acts like a "folder" or "category" for digital assets. In this case, we are creating `"Example Collection"`.

```tsx filename="index.ts"
const createCollectionTransaction = await aptos.createCollectionTransaction({
  creator: alice,
  description: "This is an example collection.",
  name: "Example Collection",
  uri: "aptos.dev",
});

const committedTxn = await aptos.signAndSubmitTransaction({
  signer: alice,
  transaction: createCollectionTransaction,
});
await aptos.waitForTransaction({ transactionHash: committedTxn.hash });
```

<Callout type="info">
Always wait for the transaction to complete using `waitForTransaction` before proceeding. This ensures the collection is ready before you try minting.
</Callout>

### Mint a Digital Asset

> With the collection created, we can now mint a digital asset (an NFT) into it. This involves providing details like the name, description, and a URI (often linking to metadata like images).

Reviewer: ```suggestion
With the collection created, we can now mint a digital asset (an NFT) for the collection. This involves providing details like the name, description, and a URI (often linking to metadata like images).
```

console.log(`Alice's digital asset balance: ${aliceDigitalAssetsAfter.length}`);
console.log(`Bob's digital asset balance: ${bobDigitalAssetsAfter.length}`);
```
</Steps>

## Running An Example

### Getting Started

<Steps>
### Prerequisites

- **Node.js**: Make sure you have Node.js (version 16 or later) and npm installed.
- **Aptos CLI (Optional)**: Having the [Aptos CLI](https://aptos.dev/cli-tools/aptos-cli-tool/install-aptos-cli) can be helpful if you want to inspect on-chain data directly, but it's not required for this tutorial.

### Set up Your Project

Create a new directory for your project and initialize a Node.js project:

```bash filename="Terminal"
mkdir aptos-digital-asset-tutorial
cd aptos-digital-asset-tutorial
npm init -y
```

This will create a `package.json` file, allowing you to install dependencies and run scripts.

### Install Dependencies

You will need the Aptos TypeScript SDK and `dotenv` to manage environment variables:

```bash filename="Terminal"
npm install @aptos-labs/ts-sdk dotenv
npm install --save-dev @types/node
```

### Create tsconfig.json

1. Create a new TypeScript configuration file:

```bash filename="Terminal"
touch tsconfig.json

> ```

Reviewer: Can we use Tabs here, one tab for mac os / linux and another for windows?

<Callout type="info">
**Windows users:** Use `type nul > tsconfig.json` in Command Prompt.
</Callout>

2. Open the `tsconfig.json` file and add the following configuration:

```json filename="tsconfig.json"
{
  "compilerOptions": {
    "target": "es2020",
    "module": "commonjs",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true,
    "types": ["node"],
    "lib": ["es2020"]
  }
}
```

<Callout type="warning">
Remember to save your file after adding the configuration!

> </Callout>

Reviewer: Nit: Remove

Reviewer 2: We did a couple user interviews and found that several folks we interviewed forgot to save and ended up confused, so that's what inspired that callout. I think it's worth keeping in the "Your First NFT" especially since NFT creators aren't necessarily technical, but may be using AI to help them get by (meaning when they miss small steps like this it's hard for them to notice why if they've already refocused on another file).

This configuration ensures TypeScript properly recognizes Node.js types and provides appropriate type checking.

### Configure Environment Variables

1. Create a `.env` file by running:

```bash filename="Terminal"
touch .env
```

<Callout type="info">
**Windows users:** Use `type nul > .env` in Command Prompt.
</Callout>

2. Open `.env` to edit it.
3. Update `.env` to have the following line:

```bash filename="Terminal"
APTOS_NETWORK=devnet
```

<Callout type="warning">
Remember to save your `.env` file after adding the configuration!
</Callout>

<Callout type="info">
By default, we'll use `devnet`, but you can also choose `testnet` or `mainnet` depending on your needs.
</Callout>

### Adding index.ts

1. Create a new file called `index.ts` in your project folder:

```bash filename="Terminal"
touch index.ts
```

<Callout type="info">
**Windows users:** Use `type nul > index.ts` in Command Prompt.

> </Callout>

Reviewer: Can we use Tabs here? one for mac / linux and another for windows?

Author: Great idea, I've just updated and replaced three sections with the Tabs component:

- In the "Create tsconfig.json" section
- In the "Configure Environment Variables" section
- In the "Adding index.ts" section

        transaction: transferTransaction,
    });

    console.log("Waiting for the transfer transaction to complete...");
    await aptos.waitForTransaction({ transactionHash: transferTxn.hash });
    console.log("Digital asset transferred successfully!");

    console.log("\n=== Step 6: Verifying digital asset balances ===\n");
    const aliceDigitalAssetsAfter = await aptos.getOwnedDigitalAssets({
        ownerAddress: alice.accountAddress,
    });
    const bobDigitalAssetsAfter = await aptos.getOwnedDigitalAssets({
        ownerAddress: bob.accountAddress,
    });

    console.log(`Alice's digital asset balance: ${aliceDigitalAssetsAfter.length}`);
    console.log(`Bob's digital asset balance: ${bobDigitalAssetsAfter.length}`);
    
    console.log("\n=== Step 7: Transaction hashes for explorer ===\n");
    console.log(`Collection creation transaction: ${committedTxn.hash}`);
    console.log(`Mint transaction: ${mintTxn.hash}`);
    console.log(`Transfer transaction: ${transferTxn.hash}`);
    console.log("\nYou can view these transactions on the Aptos Explorer:");
    console.log("https://explorer.aptoslabs.com/?network=devnet");
}

example();
```

### Run the code

> Run the code with:

Reviewer: nit: remove, already have a header titled `### Run the code`

Author: Done [x]

```bash filename="Terminal"
npx ts-node index.ts
```

<Callout type="warning">
If you encounter any errors, make sure all your files are saved and that you've followed all the steps correctly.

> </Callout>

Reviewer: Nit: rm not necessary

   - The exact time the transaction was processed
   - Gas fees paid
   - The digital asset that was transferred

This is a great way to verify your transactions and understand how they're recorded on the blockchain.
</Steps>

## Summary of the Key Steps

1. **Setup Your Project**: Initialize a Node.js project and install `@aptos-labs/ts-sdk` and `dotenv`.
2. **Configure Your Environment**: Set `APTOS_NETWORK` in a `.env` file.
3. **Configure TypeScript**: Create a proper `tsconfig.json` file to ensure type safety.
4. **Initialize the `Aptos` Client**: Connect to Aptos using `new Aptos(new AptosConfig({ network }))`.
5. **Create & Fund Accounts**: Generate and fund accounts on devnet to pay for transactions.
6. **Create a Collection:** All Digital Assets need a collection to own them.
7. **Mint a Digital Asset**: Create custom Digital Assets which point to whichever `uri` you choose.
8. **Transfer the Asset**: Move the digital asset from one account to another.
9. **Verify the Result**: Confirm that the recipient now holds the newly minted digital asset.
10. **Log Transaction Hashes**: Output all transaction hashes for easy reference and exploration.

> 11. **Explore Transactions**: View your transactions on the Aptos Explorer to understand how they're recorded on-chain.

Reviewer: Maybe remove or move to the top? The sidebar already has a table of contents and I'm not sure why a summary would be at the end.

## Troubleshooting

### Common TypeScript Errors

- **"Cannot find name 'process'"**: This error occurs when TypeScript can't find the Node.js type definitions. Make sure you have both installed `@types/node` and created a proper `tsconfig.json` file as shown in the "Configure TypeScript" step.

- **"Object is possibly 'undefined'"**: TypeScript's strict mode prevents accessing properties on potentially undefined objects. Always check if arrays have elements before accessing them, as shown in our code example.

> 

Reviewer: Nit: rm, out of scope

### Environment Variables

If you're having issues with environment variables not being recognized:
1. Make sure your `.env` file is in the root directory of your project
2. Verify the file contains `APTOS_NETWORK=devnet` (or your chosen network)
3. Check that you've imported `dotenv/config` at the top of your script

### Further Reading & Resources

> - **Aptos TypeScript SDK Reference**: [Aptos Dev Documentation](https://aptos.dev/en/build/sdks/ts-sdk)

Reviewer: ```suggestion
- [aptos ts-sdk docs](https://aptos.dev/en/build/sdks/ts-sdk)
```
